### PR TITLE
Try to detect borked package installations.

### PR DESCRIPTION
**Pull request by DMRobertson**
As in #12223.

### CODE REVIEW

File: synapse/util/check_dependencies.py [hunk 1]
```diff
@@ -128,6 +128,21 @@ def _incorrect_version(
         )
 
 
+def _no_reported_version(
+    requirement: Requirement, extra: Optional[str] = None
+):
+    if extra:
+        return (
+            f"Synapse {VERSION} needs {requirement} for {extra}, "
+            f"but can't determine {requirement.name}'s version."
```
Comment (squahtx) [inline synapse/util/check_dependencies.py:137]: consistency nit: we don't include a `.` at the end of the other messages
```suggestion
            f"but can't determine {requirement.name}'s version"
```

File: synapse/util/check_dependencies.py [hunk 2]
```diff
@@ -128,6 +128,21 @@ def _incorrect_version(
...
+        )
+    else:
+        return (
+            f"Synapse {VERSION} needs {requirement},"
```
Comment (squahtx) [inline synapse/util/check_dependencies.py:141]: nit: the message is missing a space
```suggestion
            f"Synapse {VERSION} needs {requirement}, "
```

File: tests/util/test_check_dependencies.py [hunk 3]
```diff
@@ -59,13 +60,28 @@ def test_mandatory_dependency(self) -> None:
...
+                check_requirements()
+                check_requirements("cool-extra")
             with self.mock_installed_package(old):
-                self.assertRaises(DependencyException, check_requirements)
+                check_requirements()
+                check_requirements("cool-extra")
             with self.mock_installed_package(new):
-                # should not raise
                 check_requirements()
+                check_requirements("cool-extra")
```
Comment (squahtx) [inline tests/util/test_check_dependencies.py:84]: I'm very confused by these test changes.

Why did we change `test_mandatory_dependency`? And shouldn't `test_version_reported_as_None` be the test that uses `distribution_with_no_version`?

Comment (DMRobertson) [reply tests/util/test_check_dependencies.py:84]: Excellent questions. I'm also very confused.

I think I've accidentally transposed the bodies of `test_version_reported_as_None` and `test_mandatory_dependency`. I was probably rushing :(

Comment (DMRobertson) [reply tests/util/test_check_dependencies.py]: Apologies again. [83ef7a9](https://github.com/matrix-org/synapse/pull/12244/commits/83ef7a9101cb4f4e30f05c8f2b85d8e4574d2ccf) should hopefully restore this to something sensible.

File: tests/util/test_check_dependencies.py [hunk 4]
```diff
@@ -67,6 +68,18 @@ def test_mandatory_dependency(self) -> None:
                 # should not raise
                 check_requirements()
 
+    def test_version_reported_as_None(self) -> None:
```
Comment (squahtx) [inline tests/util/test_check_dependencies.py:71]: naming convention suggestion/nit: the capital letter is unusual!

```suggestion
    def test_version_reported_as_none(self) -> None:
```

**Review by squahtx — APPROVED** [review]
lgtm!

mypy seems unhappy about something

**Comment by DMRobertson** [comment]
Sorry for the back and forth on this @squahtx. Mind giving it another sanity check?

**Pull request merged**